Add Repository details to PipelineRun Details page

abhinandan13jan opened the pull request:
# Addresses
https://issues.redhat.com/browse/ODC-6168

# ScreenShot
![Screenshot from 2021-07-29 17-59-31](https://user-images.githubusercontent.com/24852534/127500375-a86e3178-648e-4e10-80ef-8a496c56aa00.png)

![Screenshot from 2021-07-29 17-59-34](https://user-images.githubusercontent.com/24852534/127500381-b82fc4c6-f6e7-44cb-a9f6-73bb2bb7ede3.png)

# Tests
Added test Suite for PipelineRepositories

# Browser Conformance
Chrome

karthikjeeyar commented on frontend/packages/pipelines-plugin/src/components/repository/conts.ts:
```diff
@@ -0,0 +1,23 @@
+}
+
+export const repositoryLabels: Record<REPOSITORY_FIELDS, RepositoryLabel> = {
+  [REPOSITORY_FIELDS.REPOSITORY]: { labels: ['pipelinesascode.tekton.dev/repository'] },
```
Save all the required repository labels in an enum `RepositoryLabels` 

```
export enum RepositoryLabels {
    Repository = 'pipelinesascode.tekton.dev/repository',
    Branch = 'pipelinesascode.tekton.dev/branch',
    RepoUrl = 'pipelinesascode.tekton.dev/url-repository', 
    RepoOrg = 'pipelinesascode.tekton.dev/url-org'
    Sha = 'pipelinesascode.tekton.dev/sha',
    EventType = 'pipelinesascode.tekton.dev/event-type',
}
```

abhinandan13jan replied:
I am now using a single data structure, but I dont want to use string based keys as these would be used in multiple places

karthikjeeyar commented on frontend/packages/pipelines-plugin/src/components/repository/conts.ts:
```diff
@@ -0,0 +1,23 @@
+  [REPOSITORY_FIELDS.BRANCH]: { labels: ['pipelinesascode.tekton.dev/branch'] },
+  [REPOSITORY_FIELDS.REPO_URL]: {
+    labels: ['pipelinesascode.tekton.dev/url-org', 'pipelinesascode.tekton.dev/url-repository'],
+    baseURL: 'https://github.com',
```
Move this out as a separate enum called `SupportedVCS` and add `github` in there

abhinandan13jan replied:
moved out baseURL

karthikjeeyar commented on frontend/packages/pipelines-plugin/src/components/repository/PipelineRunRepositories.tsx:
```diff
@@ -0,0 +1,41 @@
+      {Object.keys(repositoryLabels).map((repositoryLabel) => {
+        const labelKey = repositoryLabels[repositoryLabel].labels;
+        if (repositoryLabel === REPOSITORY_FIELDS.REPO_URL) {
+          return plrLabels[labelKey[0]] && plrLabels[labelKey[1]] ? (
```
We could explicitly use the label here instead of using `labelKey[0]` and `labelKey[1]`, It will improve the readability.

abhinandan13jan replied:
updated

karthikjeeyar commented on frontend/packages/pipelines-plugin/src/components/repository/PipelineRunRepositories.tsx:
```diff
@@ -0,0 +1,41 @@
+import * as React from 'react';
+import { useTranslation } from 'react-i18next';
+import { PipelineRunKind } from '../../types';
+import { REPOSITORY_FIELDS, repositoryLabels } from './conts';
```
Rename the file
```suggestion
import { REPOSITORY_FIELDS, repositoryLabels } from './const';
```

abhinandan13jan replied:
renamed to consts

karthikjeeyar commented on frontend/packages/pipelines-plugin/src/components/repository/PipelineRunRepositories.tsx:
```diff
@@ -0,0 +1,41 @@
+        }
+        return plrLabels[labelKey[0]] ? (
+          <>
+            <dt>{t(repositoryLabel)}</dt>
```
Since you are looping through the object this key is dynamic value, and you cannot translate a dynamic value like this.

abhinandan13jan replied:
Using static translations now to better align

karthikjeeyar commented on frontend/packages/pipelines-plugin/src/components/repository/conts.ts:
```diff
@@ -0,0 +1,23 @@
+  baseURL?: string;
+}
+
+export enum REPOSITORY_FIELDS {
```
use PascalCase here for the enums.

abhinandan13jan replied:
updated

karthikjeeyar commented on frontend/packages/pipelines-plugin/src/components/pipelineruns/detail-page-tabs/PipelineRunCustomDetails.tsx:
```diff
@@ -57,6 +58,7 @@ const PipelineRunCustomDetails: React.FC<PipelineRunCustomDetailsProps> = ({ pip
         namespace={pipelineRun.metadata.namespace}
         ownerResourceName={pipelineRun.metadata.name}
       />
+      <PipelineRunRepositories pipelineRun={pipelineRun} />
```
A pipeline Run can associate back to only one repository, the name here needs to be changed. maybe `RepositoryLinkList` ?

abhinandan13jan replied:
renamed

karthikjeeyar reviewed (commented):
Repository link has to be a resource link and also the Repo Url needs to link to github. Refer HLD and need to adjust in few places.

karthikjeeyar commented on frontend/packages/pipelines-plugin/src/models/pipelines.ts:
```diff
@@ -205,3 +205,22 @@ export const EventListenerModel: K8sKind = {
+  label: 'Repository',
+  // t('pipelines-plugin~Repository')
+  labelKey: 'pipelines-plugin~Repository',
+  // t('pipelines-plugin~Repository')
```
```suggestion
  // t('pipelines-plugin~Repositories')
```

abhinandan13jan replied:
updated

karthikjeeyar commented on frontend/packages/pipelines-plugin/src/models/pipelines.ts:
```diff
@@ -205,3 +205,22 @@ export const EventListenerModel: K8sKind = {
+  namespaced: true,
+  kind: 'Repository',
+  id: 'repository',
+  labelPlural: 'reposities',
```
typo
```suggestion
  labelPlural: 'repositories',
```

abhinandan13jan replied:
updated

karthikjeeyar commented on frontend/packages/pipelines-plugin/src/components/repository/RepositoryLinkList.tsx:
```diff
@@ -0,0 +1,62 @@
+import * as React from 'react';
+import { GithubIcon } from '@patternfly/react-icons';
+import { useTranslation } from 'react-i18next';
```
nit: sort imports

abhinandan13jan replied:
I started off as that but it  gives a lint error
![Screenshot from 2021-08-03 14-00-14](https://user-images.githubusercontent.com/24852534/127987696-49af3944-ce85-4598-b103-0f5de78e8269.png)

karthikjeeyar reviewed (commented):
Thanks @abhinandan13jan, Code looks good to me, just few nits 


@bgliwa01 Repository CR calls it as `Event type`, shall we do the same instead of having it as `Trigger type`?

karthikjeeyar commented:
/lgtm

karthikjeeyar commented:
/assign @andrewballantyne

karthikjeeyar commented:
@abhinandan13jan You need a rebase

karthikjeeyar commented:
/lgtm cancel
Based on yesterday's ODC pipelines meeting,

1. `Trigger type` should be changed to `Event type`
2. If `SHA` value in the label is url then show the ExternalLink or it needs to be in copy to clipboard component.

abhinandan13jan commented:
> /lgtm cancel
> Based on yesterday's ODC pipelines meeting,
> 
> 1. `Trigger type` should be changed to `Event type`
> 2. If `SHA` value in the label is url then show the ExternalLink or it needs to be in copy to clipboard component.
> 
> Raised an issue in PAC to support sha as ful url [openshift-pipelines/pipelines-as-code#192](https://github.com/openshift-pipelines/pipelines-as-code/issues/192)

@karthikjeeyar 
Rebased and updated
![Screenshot from 2021-08-04 17-03-20](https://user-images.githubusercontent.com/24852534/128175118-7438ad1a-219d-4b89-b2e6-3179f3df8529.png)

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/repository/RepositoryLinkList.tsx:
```diff
@@ -0,0 +1,64 @@
+      )}
+      {plrLabels[RepositoryLabels[RepositoryFields.SHA]] && (
+        <>
+          <dt>{t('pipelines-plugin~SHA')}</dt>
```
I think we want this to be `Commit Id` cc @bgliwa01 @karthikjeeyar 

SHA is the type of format the commit id is in, it's more useful to reference it as commit id as THAT's what it is. SHA is a format and can represent anything (not necessarily a commit id)

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/repository/RepositoryLinkList.tsx:
```diff
@@ -0,0 +1,64 @@
+  const { t } = useTranslation();
+  const plrLabels = pipelineRun.metadata.labels;
+  const repoLabel = RepositoryLabels[RepositoryFields.REPOSITORY];
+  const repoName = plrLabels && plrLabels[repoLabel];
```
```suggestion
  const repoName = plrLabels?.[repoLabel];
```

abhinandan13jan replied:
updated

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/repository/__tests__/RepositoryLinkList.spec.tsx:
```diff
@@ -0,0 +1,88 @@
+import { PipeLineRunWithRepoMetadata } from '../../../test-data/pipeline-data';
+import RepositoryLinkList from '../RepositoryLinkList';
+
+jest.mock('react-i18next', () => {
+  const reactI18next = require.requireActual('react-i18next');
+  return {
+    ...reactI18next,
+    useTranslation: () => ({ t: (key: string) => key }),
+  };
+});
```
This should be already [globally mocked](https://github.com/openshift/console/blob/master/frontend/__mocks__/react-i18next.js) for you. If you _need_ this to avoid errors than our global mock is in error and that's something we should look into.

abhinandan13jan replied:
updated

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/repository/__tests__/RepositoryLinkList.spec.tsx:
```diff
@@ -0,0 +1,88 @@
+    expect(repositoryWrapper.find('dl').exists()).toBe(false);
+  });
+
+  it('should return correct details when repository related labels exists', () => {
```
I feel like we shouldn't write tests like this... Digging into components to see if a passed prop is equal to the passed value is giving us what exactly? I feel like this is just saying "we wrote that line of code to do that static passing". 

If you're concerned about a computation or logical split of code out of something, that's where utilities and hooks come into play -- test THOSE.

Also primarily we should only be testing conditional renderings... static renderings don't give us a lot. If you need to select a component to see if something is inside it, add a `data-test` attribute so it can also be used by functional tests. We should avoid selecting `2nd div, is span inside it?` kinda tests.

abhinandan13jan replied:
@andrewballantyne 
It's more around tracking that SHA is wrapped in CopyToClipboard and URL is exposed as an externalLink and Repo as a resourceLink. Also logically it checks the number of Repository related fields rendered which are conditional.

Let me whether we should keep these or remove

andrewballantyne replied:
I've asked @karthikjeeyar to follow up with you. Look out for his message or reach out to him first.

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/repository/consts.ts:
```diff
@@ -0,0 +1,19 @@
+  BRANCH = 'Branch',
+  URL_REPO = 'RepoUrl',
+  URL_ORG = 'RepoOrg',
+  SHA = 'sha',
```
This is fine being called SHA as it's the technical id of the label. It however doesn't need to be echo'ed to the user as such. (this is a comment continuation of a previous comment -- nothing to do here)

karthikjeeyar commented on frontend/packages/pipelines-plugin/src/components/repository/__tests__/RepositoryLinkList.spec.tsx:
```diff
@@ -0,0 +1,32 @@
+    expect(repoData.length).toBe(4);
+  });
+
+  it('should only the required data', () => {
+    const repositoryWrapper = shallow(
+      <RepositoryLinkList pipelineRun={PipeLineRunWithRepoMetadata[2]} />,
+    );
+    expect(repositoryWrapper.find('dl').exists()).toBe(true);
+    const repoData = repositoryWrapper.find('dl').find('dd');
+    expect(repoData.length).toBe(2);
+  });
+});
```
In this test, If the `dl` or `dd` is changed to some other component in future, then we will have a broken test.

Few things to keep in my mind while writing unit test cases 

1. Avoid using the components directly to find and assert. 
2. If there is need to test the fields existence, then it must have a data-test attribute and use that attribute to find the field rather than directly referencing the element like `'find(dl)'`
3. Add test cases for all the conditionally renderings (in your case its all the repo labels).

eg: If Branch is set in the label, then expect the Branch Field is shown in the UI
```
 {plrLabels[RepositoryLabels[RepositoryFields.BRANCH]] && (
        <>
          <dt  data-test="branch-title">{t('pipelines-plugin~Branch')}</dt>
          <dd>{plrLabels[RepositoryLabels[RepositoryFields.BRANCH]]}</dd>
        <>
      )}
  ```

abhinandan13jan replied:
I've updated with individual data-test-id's. Thanks for the review once again :)

karthikjeeyar replied:
It should be data-test instead of data-test-id, that is what we follow for all the e2e tests also
Ref: https://github.com/openshift/console/blob/master/frontend/packages/integration-tests-cypress/support/selectors.ts#L47-L50

abhinandan13jan replied:
pushed

karthikjeeyar commented:
Thanks @abhinandan13jan 
/lgtm

andrewballantyne reviewed (approved):
/retest

abhinandan13jan commented:
thanks for the reviews @karthikjeeyar @andrewballantyne

andrewballantyne commented:
/label qe-approved
/label docs-approved
/label cx-approved

andrewballantyne commented:
/label px-approved
